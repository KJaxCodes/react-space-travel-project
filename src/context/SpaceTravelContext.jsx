import React from 'react';
import { createContext, useContext, useState } from 'react';
import SpaceTravelApi from '../services/SpaceTravelApi';

const SpaceTravelContext = createContext();

export const useSpaceTravel = () => useContext(SpaceTravelContext);

export const SpaceTravelProvider = ({ children }) => {
    const [spacecrafts, setSpacecrafts] = useState([]);
    const [planets, setPlanets] = useState([]);
    const [isLoading, setIsLoading] = useState(false);


    //Actions

    //Decommission spacecraft by ID
    const decommissionSpacecraftById = async (id) => {

        try {
            setIsLoading(true);
            const response = await SpaceTravelApi.destroySpacecraftById({ id: id });
            if (response.isError) {
                throw new Error("Failed to decommission.");
            }
            setSpacecrafts((prev) => prev.filter((craft) => craft.id !== id));
        } catch (err) {
            console.error("Decommission failure", err.message);
        } finally {
            console.log("This will always run");
            setIsLoading(false);
            console.log("And we are done loading");
        }
    };

    //Assign or reassign spacecraft to planet
    const assignSpacecraftToPlanet = async ({ spacecraftId, targetPlanetId }) => {
        try {
            const response = await SpaceTravelApi.sendSpacecraftToPlanet({
                spacecraftId, targetPlanetId
            });
            if (!response.isError) {
                setSpacecrafts((prev) =>
                    prev.map((craft) =>
                        craft.id === spacecraftId ? { ...craft, currentLocation: targetPlanetId } : craft
                    )
                );
            }
        } catch (err) {
            console.error("Failed to assign spacecraft", err.message);
        }
    }

    // Load initial spacecrafts and planets

    const loadInitialData = async () => {
        try {
            setIsLoading(true);

            const spacecraftRes = await SpaceTravelApi.getSpacecrafts();
            const planetRes = await SpaceTravelApi.getPlanets();

            if (spacecraftRes.isError) {
                console.log("loadInitialData Error: Spacecraft");
                console.debug(spacecraftRes);
            }
            if (planetRes.isError) {
                console.log("loadInitialData Error: Planet");
                console.debug(spacecraftRes);
            }

            setSpacecrafts(spacecraftRes.data);
            setPlanets(planetRes.data);

        } catch (err) {
            console.error("Failed to load initial data", err.message);
        } finally {
            setIsLoading(false);
        }
    }

    return (
        <SpaceTravelContext.Provider
            value={{
                spacecrafts,
                setSpacecrafts,
                planets,
                setPlanets,
                decommissionSpacecraftById,
                assignSpacecraftToPlanet,
                loadInitialData,
                isLoading,
                setIsLoading,
            }}
        >
            {children}
        </SpaceTravelContext.Provider>
    );
};