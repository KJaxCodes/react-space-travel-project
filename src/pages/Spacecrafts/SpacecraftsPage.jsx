//Show all spacecraft with ability to click for more info 
//Display spacecraft name, capacity, current planet location 
//Option to construct a new spacecraft
//Option to delete/decommission existing spacecraft - should I have my delete button here? -refresh page? set to global state?
//Option to transfer spacecraft to a planet
//have a dropdown which would render all planets
//choose the planet to assign to craft to
//set spacecraft to the selected planet
//have a button called assign
//assigns it by target planet id and spacecraft id
import React from "react";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { useSpaceTravel } from "../../context/SpaceTravelContext";
import Loader from "../../components/Loader";

function SpacecraftsPage() {

    const {
        spacecrafts,
        decommissionSpacecraftById,
        assignSpacecraftToPlanet,
        loadInitialData,
        planets,
        isLoading,
    } = useSpaceTravel();

    //local state
    // const [isLoading, setIsLoading] = useState(true); moved to global state
    const [error, setError] = useState(null);
    const [selectedPlanetIds, setSelectedPlanetIds] = useState({});

    useEffect(() => {
        //this is all we need
        loadInitialData();

    }, []);

    const handlePlanetSelect = (craftId, planetId) => {
        setSelectedPlanetIds((prev) => ({ ...prev, [craftId]: planetId, }));
    };

    const getPlanetNameById = (planetId) => {
        const planet = planets.find((p) => p.id === planetId);
        return planet ? planet.name : "Unassigned";
    };

    if (!spacecrafts) return <p>Loading spacecrafts......</p>

    return (
        <div>
            <h2>🚀 All Spacecrafts</h2>
            <p>
                <Link to="/spacecrafts/build">🛠️ Build New Spacecraft</Link>
            </p>
            <ul>
                {
                    isLoading ? (
                        <li>
                            <Loader />
                        </li>
                    ) :
                        spacecrafts.map((craft) => (
                            <li data-testid="spacecrafts-page" key={craft.id}>
                                <Link to={`/spacecrafts/${craft.id}`} title="View details">
                                    🔍
                                </Link> {" "}
                                {craft.name}
                                <p>Capacity: {craft.capacity}</p>
                                <p>Currently on Planet: {getPlanetNameById(parseInt(craft.currentLocation))}</p>
                                <p>
                                    <label>Decommission {craft.name}:</label>
                                    <button
                                        onClick={() => decommissionSpacecraftById(craft.id)}
                                        style={{ marginLeft: "1rem" }}
                                    >
                                        Decommission
                                    </button>
                                </p>

                                <p>
                                    <label> Reassign {craft.name}:</label>{" "}
                                    <select
                                        value={selectedPlanetIds[craft.id]}
                                        onChange={(e) => handlePlanetSelect(craft.id, e.target.value)}
                                    >
                                        <option value=""> Select Planet Reassignment</option>
                                        {planets.map((planet) => (
                                            <option key={planet.id} value={planet.id}> {planet.name} </option>
                                        ))}
                                    </select> {" "}
                                    <button
                                        onClick={async () => {
                                            assignSpacecraftToPlanet({
                                                spacecraftId: craft.id,
                                                targetPlanetId: selectedPlanetIds[craft.id],
                                            });
                                            await loadInitialData();
                                        }}
                                        disabled={!selectedPlanetIds[craft.id]}
                                    >
                                        Assign
                                    </button>
                                </p>

                            </li>
                        ))
                }
            </ul>
        </div >
    );
}

export default SpacecraftsPage;

// function SpacecraftsPage() {
//     return (
//         <div>
//             <h2>🚀 All Spacecrafts</h2>
//             <p>List of all spacecraft will go here.</p>
//         </div>
//     );
// }

// export default SpacecraftsPage;