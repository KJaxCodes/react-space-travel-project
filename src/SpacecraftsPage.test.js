import React from "react";
import { render } from "@testing-library/react";

//components
import SpacecraftsPage from "./pages/Spacecrafts/SpacecraftsPage";
import { SpaceTravelProvider } from "./context/SpaceTravelContext";
import SpaceTravelApi from "./services/SpaceTravelApi";
import { MemoryRouter } from "react-router-dom";

test("Spacecrafts Render Correctly", () => {

    const { asFragment } = render(
        <MemoryRouter>
            <SpaceTravelProvider>
                <SpacecraftsPage />
            </SpaceTravelProvider>
        </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot("spacecrafts-page");
});