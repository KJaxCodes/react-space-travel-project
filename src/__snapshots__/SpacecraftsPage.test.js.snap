// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Spacecrafts Render Correctly: spacecrafts-page 1`] = `
<DocumentFragment>
  <div>
    <h2>
      🚀 All Spacecrafts
    </h2>
    <p>
      <a
        href="/spacecrafts/build"
      >
        🛠️ Build New Spacecraft
      </a>
    </p>
    <ul>
      <li>
        <p>
          ⏳Loading...⌛
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
